File: test_expectimax_aligned.py
#!/usr/bin/env python
"""Test alignment between Python and OCaml expectimax implementations."""

import subprocess
from game import Game2048
from expectimax_python import max_value, expect_value, get_best_move

def test_basic_expectimax():
    """Test basic expectimax operations."""
    Game2048._init_tables()
    
    # Simple evaluation function - just count empty cells
    def simple_eval(board):
        empty_count = sum(1 for i in range(16) if ((board >> (4*i)) & 0xF) == 0)
        return float(empty_count)
    
    test_boards = [
        0x0000000000001234,  # Simple board
        0x1234567890ABCDEF,  # Complex board
        0x0000000000000000,  # Empty board
    ]
    
    print("Testing Python Expectimax Implementation")
    print("=" * 50)
    
    for board in test_boards:
        print(f"\nBoard: {board:016x}")
        
        # Test max_value at different depths
        for depth in [0, 1, 2]:
            value = max_value(board, simple_eval, depth)
            print(f"  max_value(depth={depth}): {value:.2f}")
        
        # Test best move
        move, score = get_best_move(board, simple_eval, 2)
        print(f"  best_move(depth=2): {move} (score: {score:.2f})")

def test_deterministic_game():
    """Test a deterministic game sequence."""
    Game2048._init_tables()
    
    print("\n\nTesting Deterministic Game Sequence")
    print("=" * 50)
    
    # Use a more sophisticated evaluation
    def eval_func(board):
        empty_count = sum(1 for i in range(16) if ((board >> (4*i)) & 0xF) == 0)
        max_tile = 0
        for i in range(16):
            tile = (board >> (4*i)) & 0xF
            if tile > max_tile:
                max_tile = tile
        return float(empty_count) + float(max_tile) * 10.0
    
    # Start with a specific board
    board = 0x0000000000001122  # [2,2,1,1] in bottom row
    print(f"Initial board: {board:016x}")
    
    # Make a few moves
    for i in range(3):
        move, score = get_best_move(board, eval_func, 2)
        if move is None:
            print("No valid moves!")
            break
            
        print(f"\nMove {i+1}: {move} (expected score: {score:.2f})")
        
        # Execute the move
        if move == 'up':
            board, move_score, _ = Game2048.move_up(board)
        elif move == 'down':
            board, move_score, _ = Game2048.move_down(board)
        elif move == 'left':
            board, move_score, _ = Game2048.move_left(board)
        elif move == 'right':
            board, move_score, _ = Game2048.move_right(board)
        
        print(f"After move: {board:016x} (actual score: {move_score})")

if __name__ == "__main__":
    test_basic_expectimax()
    test_deterministic_game()

File: test_expectimax_aligned.ml
(* Test alignment of expectimax implementation *)

open Gp_2048_lib.Game
open Gp_2048_lib.Gp_tree
open Gp_2048_lib.Expectimax_aligned

(* Simple evaluation function - count empty cells *)
let _simple_eval board =
  float_of_int (count_empty_cells board)

(* Create a simple program that just counts empty cells *)
let simple_program = {
  nodes = [| NumEmptyCells |];
  fitness = 0.0;
  games_played = 0;
  avg_score = 0.0;
  avg_max_tile = 0.0;
}

let test_basic_expectimax () =
  Printf.printf "Testing OCaml Expectimax Implementation\n";
  Printf.printf "==================================================\n";
  
  let test_boards = [
    Int64.of_string "0x0000000000001234";  (* Simple board *)
    Int64.of_string "0x1234567890ABCDEF";  (* Complex board *)
    Int64.of_string "0x0000000000000000";  (* Empty board *)
  ] in
  
  List.iter (fun board ->
    Printf.printf "\nBoard: %016Lx\n" board;
    
    (* Test max_value at different depths *)
    List.iter (fun depth ->
      let value = max_value board simple_program depth in
      Printf.printf "  max_value(depth=%d): %.2f\n" depth value
    ) [0; 1; 2];
    
    (* Test best move *)
    match get_best_move board simple_program 2 with
    | None -> Printf.printf "  best_move(depth=2): None\n"
    | Some dir ->
      let dir_str = match dir with
        | `Up -> "up"
        | `Down -> "down"
        | `Left -> "left"
        | `Right -> "right"
      in
      (* Calculate the score by evaluating the move *)
      let new_board = match dir with
        | `Up -> move_up board
        | `Down -> move_down board
        | `Left -> move_left board
        | `Right -> move_right board
      in
      let score = expect_value new_board simple_program 2 in
      Printf.printf "  best_move(depth=2): %s (score: %.2f)\n" dir_str score
  ) test_boards

let test_deterministic_game () =
  Printf.printf "\n\nTesting Deterministic Game Sequence\n";
  Printf.printf "==================================================\n";
  
  (* More sophisticated evaluation *)
  let eval_program = {
    nodes = [| Add; Mul; NumEmptyCells; Constant 10.0; MaxTileValue |];
    fitness = 0.0;
    games_played = 0;
    avg_score = 0.0;
    avg_max_tile = 0.0;
  } in
  
  (* Start with a specific board *)
  let initial_board = Int64.of_string "0x0000000000001122" in
  Printf.printf "Initial board: %016Lx\n" initial_board;
  
  (* Make a few moves *)
  let board = ref initial_board in
  for i = 1 to 3 do
    match get_best_move !board eval_program 2 with
    | None -> 
      Printf.printf "No valid moves!\n";
      ()
    | Some dir ->
      let dir_str = match dir with
        | `Up -> "up"
        | `Down -> "down"
        | `Left -> "left"
        | `Right -> "right"
      in
      
      (* Calculate expected score *)
      let new_board = match dir with
        | `Up -> move_up !board
        | `Down -> move_down !board
        | `Left -> move_left !board
        | `Right -> move_right !board
      in
      let expected_score = expect_value new_board eval_program 2 in
      Printf.printf "\nMove %d: %s (expected score: %.2f)\n" i dir_str expected_score;
      
      (* Execute the move *)
      let move_score = get_score_for_move !board dir in
      board := new_board;
      Printf.printf "After move: %016Lx (actual score: %d)\n" !board move_score
  done

let () =
  test_basic_expectimax ();
  test_deterministic_game ()